lib/internal/quic/core.js

```diff
@@ -0,0 +1,3343 @@
+  // enabled. For client QuicSessions, true only if session
+  // resumption is used and early data was accepted during
+  // the TLS handshake. The value is set only after the
+  // TLS handshake is completed (immeditely before the
+  // secure event is emitted)
+  get usingEarlyData() {
+    return this.#earlyData;
+  }
+
+  get maxStreams() {
+    let bidi = 0;
+    let uni = 0;
+    if (this[kHandle]) {
+      bidi = this[kHandle].state[IDX_QUIC_SESSION_STATE_MAX_STREAMS_BIDI];
+      uni = this[kHandle].state[IDX_QUIC_SESSION_STATE_MAX_STREAMS_UNI];
+    }
+    return { bidi, uni };
+  }
+
+  get address() {
+    return this.#socket ? this.#socket.address : {};
+  }
+
+  get maxDataLeft() {
+    return this[kHandle] ?
+      this[kHandle].state[IDX_QUIC_SESSION_STATE_MAX_DATA_LEFT] : 0;
+  }
+
+  get bytesInFlight() {
+    return this[kHandle] ?
+      this[kHandle].state[IDX_QUIC_SESSION_STATE_BYTES_IN_FLIGHT] : 0;
+  }
+
+  get blockCount() {
+    return this[kHandle] ?
+      this[kHandle].state[IDX_QUIC_SESSION_STATS_BLOCK_COUNT] : 0;
+  }
+
+  get authenticated() {
+    // Specifically check for null. Undefined means the check has not
+    // been performed yet, another other value other than null means
+    // there was an error
+    return this.#verifyErrorReason == null;
+  }
+
+  get authenticationError() {
+    if (this.authenticated)
+      return undefined;
+    // eslint-disable-next-line no-restricted-syntax
+    const err = new Error(this.#verifyErrorReason);
+    const code = 'ERR_QUIC_VERIFY_' + this.#verifyErrorCode;
+    err.name = `Error [${code}]`;
+    err.code = code;
+    return err;
+  }
+
+  get remoteAddress() {
+    const out = {};
+    if (this[kHandle])
+      this[kHandle].getRemoteAddress(out);
+    return out;
+  }
+
+  get handshakeComplete() {
+    return this.#handshakeComplete;
+  }
+
+  get handshakeConfirmed() {
+    return Boolean(this[kHandle] ?
+      this[kHandle].state[IDX_QUIC_SESSION_STATE_HANDSHAKE_CONFIRMED] : 0);
+  }
+
+  get idleTimeout() {
+    return this.#idleTimeout;
+  }
+
+  get alpnProtocol() {
+    return this.#alpn;
+  }
+
+  get cipher() {
+    const name = this.#cipher;
+    const version = this.#cipherVersion;
+    return this.handshakeComplete ? { name, version } : {};
+  }
+
+  getCertificate() {
+    return this[kHandle] ?
+      translatePeerCertificate(this[kHandle].getCertificate() || {}) : {};
+  }
+
+  getPeerCertificate(detailed = false) {
+    return this[kHandle] ?
+      translatePeerCertificate(
+        this[kHandle].getPeerCertificate(detailed) || {}) : {};
+  }
+
+  ping() {
+    if (!this[kHandle])
+      throw new ERR_QUICSESSION_DESTROYED('ping');
+    this[kHandle].ping();
+  }
+
+  get servername() {
+    return this.#servername;
+  }
+
+  get destroyed() {
+    return this.#destroyed;
+  }
+
+  get closing() {
+    return this.#closing;
+  }
+
+  get closeCode() {
+    return {
+      code: this.#closeCode,
+      family: this.#closeFamily
+    };
+  }
+
+ 
```

Comment:
Same as the other comment, are you sure you want to make it emit `'finish'`. `this.writable = false` might be enough (i.e. it's the same thing - the `'finish'` part).